pkg/operation/botanist/controlplane/resourcemanager/resource_manager.go

```diff
@@ -0,0 +1,553 @@
+		}}
+		return nil
+	})
+	return err
+}
+
+func (r *ResourceManager) emptyRoleBindingInWatchedNamespace() (*rbacv1.RoleBinding, error) {
+	if r.config.WatchedNamespace == nil {
+		return nil, errors.New("creating RoleBinding in watched namespace failed - no namespace defined")
+	}
+	return &rbacv1.RoleBinding{ObjectMeta: metav1.ObjectMeta{Name: "gardener-resource-manager", Namespace: *r.config.WatchedNamespace}}, nil
+}
+
+func (r *ResourceManager) ensureService(ctx context.Context) error {
+	const (
+		healthPortName  = "health"
+		metricsPortName = "metrics"
+	)
+
+	service := r.emptyService()
+	_, err := controllerutil.CreateOrUpdate(ctx, r.client, service, func() error {
+		service.Labels = *r.config.DefaultLabels
+		service.Spec.Selector = *appLabel()
+		service.Spec.Type = corev1.ServiceTypeClusterIP
+		service.Spec.ClusterIP = corev1.ClusterIPNone
+		service.Spec.Ports = kutil.ReconcileServicePorts(service.Spec.Ports, []corev1.ServicePort{
+			{
+				Name:     metricsPortName,
+				Protocol: corev1.ProtocolTCP,
+				Port:     metricsPort,
+			},
+			{
+				Name:     healthPortName,
+				Protocol: corev1.ProtocolTCP,
+				Port:     healthPort,
+			},
+		})
+		return nil
+	})
+	return err
+}
+
+func (r *ResourceManager) emptyService() *corev1.Service {
+	return &corev1.Service{ObjectMeta: metav1.ObjectMeta{Name: "gardener-resource-manager", Namespace: r.namespace}}
+}
+
+func (r *ResourceManager) ensureDeployment(ctx context.Context) error {
+	const (
+		limitCPU                  = "400m"
+		limitMemory               = "512Mi"
+		requestCPU                = "23m"
+		requestMemory             = "47Mi"
+		volumeMountName           = "gardener-resource-manager"
+		volumeMountPathKubeconfig = "/etc/gardener-resource-manager"
+		volumeName                = "gardener-resource-manager"
+	)
+
+	deployment := r.emptyDeployment()
+
+	_, err := controllerutil.CreateOrUpdate(ctx, r.client, deployment, func() error {
+		deployment.Labels = *r.config.DefaultLabels
+
+		deployment.Spec.Replicas = &r.replicas
+		deployment.Spec.RevisionHistoryLimit = pointer.Int32Ptr(0)
+		deployment.Spec.Selector = &metav1.LabelSelector{MatchLabels: *appLabel()}
+
+		deployment.Spec.Template = corev1.PodTemplateSpec{
+			ObjectMeta: metav1.ObjectMeta{
+				Labels: utils.MergeStringMaps(*r.config.DefaultLabels, map[string]string{
+					v1beta1constants.LabelNetworkPolicyToDNS:            v1beta1constants.LabelNetworkPolicyAllowed,
+					v1beta1constants.LabelNetworkPolicyToShootAPIServer: v1beta1constants.LabelNetworkPolicyAllowed,
+					v1beta1constants.LabelNetworkPolicyToSeedAPIServer:  v1beta1constants.LabelNetworkPolicyAllowed,
+				}),
+			},
+			Spec: corev1.PodSpec{
+				ServiceAccountName: serviceAccountName,
+				Containers: []corev1.Container{
+					{
+						Name:            containerName,
+						Image:           r.image,
+						ImagePullPolicy: corev1.PullIfNotPresent,
+						Command:         r.printCommand(volumeMountPathKubeconfig),
```

Comment:
```suggestion
						Command:         r.printCommand(),
```

`printCommand` is only called once - we can just use the constant in the function?

Reply:
+1, nice catch. Actually, `printCommand` is not the best name because nothing is printed. What about `computeCommand` like in other components?

Reply:
@danielfoehrKn can you elaborate again? the command is computed with several if clauses. How would you put that into a constant in the function?
I am very fine with renaming it to computeCommand.

Reply:
@danielfoehrKn questions why the (now renamed) `computeCommand()` function has the `volumeMountPathKubeconfig` parameter. You only call `computeCommand()` once and the value you pass for this parameter itself is a constant. Can you please move such constants out of the function and to the top of the package? This is a) how it's done in other packages, and b) allows you to simplify the signature of the `computeCommand()` function.

Reply:
I guess it is a matter of code style. I dislike global variabels/constants. The variable is used twice, once in ensureDeployment and once in computeCommand. My thought here is to limit the scope of the variable as much as possible so I declared it in the ensureDeployment function. I value that higher than having to pass a variable to the computeCommand function.
Would you still prefer to have it declared at the top? Then I will do so as I also see value in adhering to the existing code style.

Reply:
Well, TBH: I don't care at all about this detail because the impact/value is very low/zero. I just mainly care about consistency and similarity to other "componentified" packages to make it easier to connect dots, especially for new or rare contributors. The impact/value of this is probably greater than zero.

With this argument, I'd personally prefer to do this like in other packages for now, i.e., have the constant at the top. It's still private to this package and not "global".

However, I won't insist on it because, as said, the impact is the "consequence" is very low. Also, it can still be improved (consistently for all packages) in the future if you see need, but probably our time is better invested elsewhere. :)

Reply:
Ok. I think it still holds value to explain to you my thought process because I actually reasoned about this trade-off when developing and I wanted to hear your opinions.
I follow your point that consistency in the code base is more important than stylistic choices and that a refactoring can then be done separately in another PR touching all the different files/components.
Hope it was not too annoying to argue here and I will carry what what I learned to future PRs :).

Reply:
No worries, many things are not obvious (especially not, if you are not very familiar with the code base or, particularly, those other packages), and I totally agree to your arguments. In general, it's a good guideline to not expose things too widely if not necessary. However, in those particular cases, it probably doesn't make much of a difference and has a low impact(, and probably a "larger value" when we strive for the consistency/similarity goal). Anyways, I'm still fine with whatever you prefer. And please continue to bring up similar thoughts in the future!